### DISCUSSION TITLE
use form validation with Route Attributes

### DISCUSSION THREAD

**Discussion by araeuchle** [spatie/laravel-route-attributes#37] [Q&A]
Hey, 
i want to use this excellent package for my current project, where i am currently testing out some ideas. 
What i encountered is, that this package currently does not work when using it with FormRequest Validators. 

Steps to reproduce. 

1. Fresh Laravel Installation 
2. create a model 
3. create a controller 
4. create a request object 
5. install spatie/laravel-route-attributes package 
6. create a get and a post route in the controller
7. validate data in post action request 
8. output form errors

        class RegistrationController extends Controller
        {
               #[Get('register', name: "register.view")]
               public function view()
               {
                   return response()->view('register.view');
               }
      
              #[Post('register', name: "register.store")]
              public function store(PublicRegisterRequest $request)
              {
                  exit("called");
              }
        }
      

the view looks like this:

       @extends('frontend')
      
       @section('content')
          <form action="{{ route('register.store') }}" method="post">
              @csrf
      
              @include('partials.form.elements.text', [
                  'name' => 'name',
                  'label' => __('public/register.name')
              ])
      
              @include('partials.form.elements.email', [
                 'name' => 'email',
                 'label' => __('public/register.email')
              ])
      
              @include('partials.form.elements.confirmed', [
                 'name' => 'password',
                 'label' => __('public/register.password')
              ])
      
              @include('partials.form.elements.checkbox', [
                'name' => 'agb',
                'label' => 'AGB'
              ])
      
              @include('partials.form.elements.submit', [
                'label' => __('public/register.submit'),
              ])
      
          </form>
      @endsection

and for example the text helper looks like this

      <div class="form-group">
          <label for="{{ $name }}">
              {{ $label }}
          </label>
      
          <input type="text" value="{{ old($name) }}" name="{{ $name }}" class="form-control" id="{{ $name }}" />
          @if (isset($errors) && $errors->has($name))
              {{ $errors->first($name) }}
          @endif
      </div>

The problem here is, that $errors are always not present. 
As soon as i remove the package, they will be outputted again

**Laravel Version**: 8.46.0
**PHP Version**: 8.0.7

Any hints how to get it to work? am i missing something?

EDIT:

When setting enabled to false and registrating the routes by myself in the web.php file, the validation works as expected.

EDIT 2: 

Also using $this->validate inside the controller action does not work. 
So currently with your package outputting validation errors is not possible.

```
   #[Post('register', name: "register.store")]
    public function store(Request $request)
    {
        $this->validate($request, [
            'name' => 'required|unique:users,name',
            'email' => 'required|unique:users,email|email',
            'password' => 'required|min:8|confirmed',
            'agb' => 'required'
        ]);


        exit("called");
    }
```

**Comment by araeuchle** [comment]
Ok, managed to solve it. You have to go into route-attributes.php and add the needed Middlewares to make it work. 

```
<?php

return [
    /*
     *  Automatic registration of routes will only happen if this setting is `true`
     */
    'enabled' => true,

    /*
     * Controllers in these directories that have routing attributes
     * will automatically be registered.
     */
    'directories' => [
        app_path('Http/Controllers'),
    ],

    /**
     * This middleware will be applied to all routes.
     */
    'middleware' => [
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
        \Illuminate\Session\Middleware\StartSession::class,
        \Illuminate\View\Middleware\ShareErrorsFromSession::class,
    ]
];

```

Question: is this really the way it should work? 
So basically we need to copy all the Middlewares we need from app/Http/Kernel.php and add them to this middleware array?

**Reply by StyxUA** [reply]
Aren't you supposed to define routes middleware yourself? Like `Route::group(['middleware' => ['web']], ...)` originally?

**Reply by araeuchle** [reply]
yes but in this package they use a different way to register routes via PHP 8 Attributes so i think sort of you have to put em there. 


@freekmurze  can you give me a hint if thats the way its supposed to work?

**Reply by msacharewicz** [reply]
Oldie, but deserves an explanation for posterity, because the `README.md` ain't very clear about it. There's no need to _copy_ the middleware. Just define the relevant middleware groups in the config file:

```php
'directories' => [
    app_path('Http/Controllers') => [
        'middleware' => ['web']
    ],
],
```
...and it will bring back the default controllers defined in `app/Http/Kernel.php`.

**Reply by freekmurze** [reply]
Feel free to send PRs to the readme to make it more clear.
